wxDFast/man/man1/wxdfast.1
.\" 
.TH "wxDownload Fast" "1" "0.3.0" "Max Velasques" ""
.SH "NAME"
wxDownload Fast \- a multithread download manager created using the wxWidgets library.

.SH "SYNTAX"
.B wxdfast
[options] [URL]

.TP 6
[URLS],
Urls space separeted of the files to be downloaded. 
Ex: www.foo.com/foo.exe ftp://ftp.foo.com/foo2.exe http://foo.com:8080/foo3.exe

.SH "DESCRIPTION"
The 
.B wxDownload Fast 
is a multithread download manager that has the follow characteristics:

* Faster downloads (with Segmented/Multi-threaded/Accelerated transfers)

* Download resuming (Pause and restart where you stopped)

* Download scheduling

* Organizes files you have already downloaded

* View server messages (HTTP, FTP, file://). No HTTPS support.

* Available in multiple languages and easily translated.  Now available in Portuguese [Brazil], Spanish, English and German 

* Connection to FTP servers which require a password

* Calculates the MD5 checksum of downloaded files so they can be easily verified

.SH "OPTIONS"
.TP 6
\-\-hide,
Start with the the main frame hide

.TP 6
\-\-notray,
Don't show the icon on system tray

.TP 6
\-list TEXTFILE, \-\-list=TEXTFILE,
Parse a text file with the list of files to download

.TP 6
\-destination DIR, \-\-destination=DIR,
Destination directory

.TP 6
\-reference URL, \-\-reference=URL,
Parse a URL that will be set as a reference URL

.TP 6
\-comments COMMENTS, \-\-comments=COMMENTS,
Add comments for the download

.SH "FIREFOX INTEGRATION"

To activate the Firefox integration, do the follow:

1) Install the FlashGot extension to the Firefox(www.flashgot.org/getit)

2) After restart the Firefox, go to Tools \-> FlashGot \-> More Options

3) Now select the wxDownload Fast on the Download Manager's List


wxDFast/ChangeLog
0.6.1
===================================================
- Added Italian language(by Angelo Contardi)
- Added Japanese language(by Tilt)
- Changes in configure.in to use `wx-config 
  --utility=wxrc` to get the location of wxrc rather
  than having it hardcoded (Patch by dirtyepic)
- Changes to make possible to build using MS VC2005
  (Patch by stephans)
- Added support for urls like this:
  ftp://user:password@www.server.com/file_or_folder.bin
  This URLs make possible to use different user/pass
  for each URL

0.6.0
===================================================
- Added Turkish language(by E.B)
- Added Indonesian language(by Arif Rahman)
- Added French language(by lord_nogard)
- Added Dutch language(by Rob van den Boorn)
- Now the download list is updated just when a
  download is changed, making the program faster,
  and preventing the program to be blocked when
  working with several downloads
- Added a option to add new files on top of the list
- Added support to more segments(until 32)
- Added portable mode(Windows only), which allow the
  program to run directly from pendrives
- Fixed some memory leaks related to the metalink
  class
- Added proxy support
- Added wxSHA1 class

0.5.4
===================================================
- The new RipStop theme now is complete
- Added Russian language(by Ruslan Fedyarov)
- Added Hungarian language(by Erno Szabados)
- Added Polish language(by Marcin Trybus)
- Added download status auto save every 10 seconds
- Added option to enable/disable checking for new
  releases
- Fixed problem when the current speed is bigger
  than graph scale
- Fixed several bugs

0.5.3
===================================================
- Added Czech language(by David Vachulka)
- Fixed problem when retriving some metalink files

0.5.2
===================================================
- Added a partial new theme called RipStop
- Now if you try to set as destination a directory
  that don't exists, the program ask to create it
- Fixed some flickering problems on the graph and
  the progress bar
- Added option to execute a custom command when
  some download is finished
- Updated to build with wxWidgets 2.7.1


0.5.1
===================================================
- Added support for multiples files inside metalinks
- Fixed bug when saving graph colors
- Fixed bug when getting file size on FTP servers

0.5.0
===================================================
- Fixed bug when the program access a server and 
  receive a 404 response
- Added functions to get the "Program Files" and
  "My Documents" directory
- Added disk space checking before to join the file
  parts after finish a download
- Added new configuration method based on autoconf
- Added a band width control, which permits the user
  to set a upper download rate
- Added metalink support. For more informations
  about what is metalink visit www.metalinker.org
- Added a tray icon menu to easily change the band
  configuration

0.4.5
===================================================
- Fixed bug on "Copy URL" and "Copy Download data"
- Fixed problem on draw the progress bar for down-
  loads splited in a odd number of parts
- Segmented progress bar with different colors
- Fixed bug on "New Download" box, when the user
  let the address box empty
- Added "--reference" option, see the Readme file
  for more informations
- Added option to choose where to save the temp
  files (.part)
- Added a new "About" box
- Added a "Donate" menu
- Added a routine to check if exists new releases
  of wxDownload Fast

0.4.4
===================================================
- Added multiple selections on Finished and In Pro-
  gress list
- Fixed problem on New Download box
- Now is possible to change the language on the fly
- Added multi-segmented progress bar

0.4.3
===================================================
- Fixed bug that change the first item of the
  finished list, when the user click in "Open
  destination directory"
- Fixed problem when getting the size of finished
  files bigger than 2Gb
- Fixed some memory access bug
- Made a total change on mDownloadFile class,
  making the class more encapsulated. This class
  keeps the download data.
- Double-click on In Progress list Start/Stop the
  downloads
- Now is possible to add several urls for download,
  this way you can download a file from several
  different servers.

0.4.2
===================================================
- Fixed problem with the Calendar Control on the
  schedule option
- Fixed problems with forbidden filename characters
- Fixed problems with the calendar box
- Added dialog box that ask about to open HTML
  files on the browser
- Added "Open Destination Directory" command that
  opens the destination directory of selected
  download
- Added tab "Paths" in the Options Box. In this are
  defined the external programs used, like the 
  browser, file manager, and so on
- Now the file size are compared with the already
  saved values. In this way, we have (almost) sure
  that the file size not changed since the last
  time, and that the file will be not corrupted
- The program internal structure was improved
- Added some minor features

0.4.1
===================================================
- Added German language
- Now the speed values that are bigger them the
  graph scale aren't showed
- Added status bar and "Stop All downloads" and
  "Start All downloads" buttons

0.4.0
===================================================
- Added support for FTP and HTTP downloads bigger
  than 2GB
- Fixed speed calculation bug with FTP downloads
- Improved the access for dynamic links
- Added option to hide the speed graph, and to
  change the graph height
- Now the last selected options in the New Down-
  load box are remembered for the next download
- Fixed some bugs

0.3.2
===================================================
- Removed some missed debug messages
- Fixed some bugs

0.3.1

wxDFast/src/wxDFast.cpp
//
// C++ Implementation: wxDFast
//
// Description: Implements the program initialization.
//
//
// Author: Max Magalhães Velasques <maxvelasques@gmail.com>, (C) 2006
//
// Copyright: See COPYING file that comes with this distribution
//
//

#include "wxDFast.h"
#include "wx/listimpl.cpp"
#include "wx/arrimpl.cpp"

WX_DEFINE_LIST(mDownloadListType)
WX_DEFINE_LIST(mUrlList)
WX_DEFINE_LIST(mGraphPoints)
WX_DEFINE_OBJARRAY(mListSelection)

IMPLEMENT_APP(mApplication)

wxConnectionBase *mServer::OnAcceptConnection(const wxString& topic)
{
    if ( topic == IPC_TOPIC + wxGetUserId())
    {
        mConnection *connection = new mConnection();
        return connection;
    }
    return NULL;
}

bool mConnection::OnExecute(const wxString& topic, wxChar* data, int size, wxIPCFormat format)
{
    wxString urlsparameter = data;
    wxArrayString urls;
    wxStringTokenizer *tkz01, *tkz02;
    int numberofparts, startoption,ontop;
    wxString textfile,destination,comments,command,reference,strurls,strtemp;
    wxTextFile file;

    tkz01 = new wxStringTokenizer(urlsparameter, wxT("$"));
    textfile = tkz01->GetNextToken();
    destination = tkz01->GetNextToken();
    comments = tkz01->GetNextToken();
    reference = tkz01->GetNextToken();
    strurls = tkz01->GetNextToken();
    tkz02 = new wxStringTokenizer(strurls, wxT("#"));
    while (tkz02->HasMoreTokens())
    {
        strtemp = tkz02->GetNextToken();
        if (strtemp != wxEmptyString)
            urls.Add(strtemp);
    }
    delete tkz01;
    delete tkz02;
    wxLogNull nolog;
    if ((textfile != wxT("NONE")) && (file.Open(textfile)))
    {
        for ( strtemp = file.GetFirstLine(); !file.Eof(); strtemp = file.GetNextLine() )
            if (strtemp != wxEmptyString)
                urls.Add(strtemp);
        if (strtemp != wxEmptyString)
            urls.Add(strtemp);
    }
    if (wxGetApp().mainframe->programoptions.rememberboxnewoptions)
    {
        startoption = wxGetApp().mainframe->programoptions.laststartoption;
        ontop = wxGetApp().mainframe->programoptions.lastontopoption;
        numberofparts = wxGetApp().mainframe->programoptions.lastnumberofparts;
        if (destination == wxT("NONE"))
            destination = wxGetApp().mainframe->programoptions.lastdestination;
        command = wxGetApp().mainframe->programoptions.lastcommand;
    }
    else
    {
        command = wxEmptyString;
        startoption = DEFAULT_START_OPTION;
        ontop = DEFAULT_ONTOP_OPTION;
        numberofparts = DEFAULT_NUM_PARTS;
        if (destination == wxT("NONE"))
            destination = wxGetApp().mainframe->programoptions.destination;
    }
    if (comments == wxT("NONE"))
        comments = wxEmptyString;
    if (reference == wxT("NONE"))
        reference = wxEmptyString;

    if (urls.GetCount() > 0)
        wxGetApp().mainframe->NewDownload(urls,destination,-1,numberofparts,wxEmptyString,wxEmptyString,reference,comments,command,startoption,ontop,TRUE,TRUE);
    else if (!wxGetApp().mainframe->IsShown())
    {
        wxCommandEvent event;
        wxGetApp().mainframe->taskbaricon->OnHide(event);
    }
    return TRUE;
}

bool mApplication::NewInstance()
{
    wxString server = IPC_SERVICE;
    wxString hostName;
    wxString textfile, destination, comments, reference, urlparameter = wxEmptyString,stringtosend;
    hostName = wxGetHostName();
    bool returntmp;

    mClient *client = new mClient;
    wxLogNull nolog;
    connection = (mConnection *)client->MakeConnection(hostName, server, IPC_TOPIC + wxGetUserId());
    if (!connection)
        returntmp = FALSE;
    else
    {
        if (!wxGetApp().parameters->Found(wxT("list"),&textfile))
            textfile = wxT("NONE");
        if (!wxGetApp().parameters->Found(wxT("destination"),&destination))
            destination = wxT("NONE");
        if (!wxGetApp().parameters->Found(wxT("comments"),&comments))
            comments = wxT("NONE");
        if (!wxGetApp().parameters->Found(wxT("reference"),&reference))
            reference = wxT("NONE");

        for (unsigned int i = 0; i < parameters->GetParamCount(); i++)
            urlparameter += parameters->GetParam(i) + wxT("#");
        stringtosend = textfile + wxT("$") + destination + wxT("$") + comments + wxT("$") + reference + wxT("$") + urlparameter + wxT("$");
        connection->Execute((wxChar *)stringtosend.c_str(),(stringtosend.Length() + 1) * sizeof(wxChar));
        returntmp = TRUE;
    }

    delete connection;
    delete client;
    return returntmp;
}

mApplication::mApplication(): m_condAllDone(m_mutexAllDone)
{
    // the mutex associated with a condition must be initially locked, it will
    // only be unlocked when we call Wait()
    #ifndef DISABLE_MUTEX
    m_mutexAllDone.Lock();
    #endif
    m_waitingUntilAllDone = FALSE;
    m_locale = NULL;
}

mApplication::~mApplication()
{
    delete m_locale;
    delete dummy;
    delete m_server;
    delete parameters;
}

bool mApplication::OnInit()
{
    parameters = new wxCmdLineParser(cmdlinedesc, argc, argv);
    if (parameters->Parse() != 0)
        exit(TRUE);

    if (NewInstance())
    return FALSE;

    #ifdef __WXMSW__
    //GET THE PROGRAM VOLUME
    wxFileName tempprogramvolume(argv[0]);
    programvolume = tempprogramvolume.GetVolume();
    #endif

    wxString service = IPC_SERVICE; //REGISTER THIS INSTANCE
    m_server = new mServer;
    m_server->Create(service);

    dummy = new wxSocketBase(); //TO SOCKETS TO WORK INSIDE THREADS
    dummy->SetTimeout(0);

    wxImage::AddHandler(new wxXPMHandler);
    wxImage::AddHandler(new wxPNGHandler);
    wxXmlResource::Get()->InitAllHandlers();
    wxString resourcepath = wxT("resources/");
    if (!wxFileExists(resourcepath + wxT("mainwindow.xrc")))
    {
        //THIS IS NEVER GOING TO HAPPEN ON WINDOWS SYSTEM
        /*#ifdef __WXMSW__
        wxFileName programpath(parameters->GetParam(0));
        resourcepath = programpath.GetPath(wxPATH_GET_SEPARATOR | wxPATH_GET_VOLUME) + resourcepath;
        #else*/
        resourcepath = wxT("/usr/share/wxdfast/");
        //#endif
    }
    themepath = resourcepath + wxT("RipStop/");
    #ifdef USE_EXTERNAL_XRC
    wxXmlResource::Get()->Load(resourcepath + wxT("mainwindow.xrc"));
    wxXmlResource::Get()->Load(resourcepath + wxT("menubar.xrc"));
    wxXmlResource::Get()->Load(resourcepath + wxT("toolbar.xrc"));
    wxXmlResource::Get()->Load(resourcepath + wxT("boxnew.xrc"));
    wxXmlResource::Get()->Load(resourcepath + wxT("boxoptions.xrc"));
    wxXmlResource::Get()->Load(resourcepath + wxT("boxabout.xrc"));
    #else
    InitXmlResource();
    #endif

    SetLanguage(mApplication::Configurations(READ,LANGUAGE_REG,0)); //SET THE LANGUAGE

    downloadlist.DeleteContents(TRUE);
    downloadlist.LoadDownloadListFromDisk();
    mainframe = NULL;
    mainframe = new mMainFrame();
    {
        int x,y,width,height,maximized,separatorposition01,separatorposition02,details;
        wxSplitterWindow *splitter01 = XRCCTRL(*mainframe, "splitter01",wxSplitterWindow);
        wxSplitterWindow *splitter02 = XRCCTRL(*mainframe, "splitter02",wxSplitterWindow);
        x = mApplication::Configurations(READ,POS_X_REG,-1);
        y = mApplication::Configurations(READ,POS_Y_REG,-1);
        width = mApplication::Configurations(READ,SIZE_X_REG,-1);
        height = mApplication::Configurations(READ,SIZE_Y_REG,-1);
        maximized = mApplication::Configurations(READ,MAXIMIZED_REG,0);
        separatorposition01 = mApplication::Configurations(READ,SEPARATOR01_REG,-1);
        separatorposition02 = mApplication::Configurations(READ,SEPARATOR02_REG,-1);
        details = mApplication::Configurations(READ,DETAILS_REG,-1);
        mainframe->SetSize(x,y,width,height);
        if (maximized)               mainframe->Maximize(TRUE);
        if (!details)            splitter01->Unsplit();
        if (separatorposition01  < 20)    separatorposition01 = 100;
        if (separatorposition02  < 20)    separatorposition02 = 100;
        splitter01->SetSashPosition(separatorposition01);
        splitter02->SetSashPosition(separatorposition02);
    }

    //IF A URL OR A TEXT FILE WAS PASSED BY THE COMMAND LINE
    int startoption, numberofparts,ontop;
    wxString listtextfile,reference,comments,destination,command;
    wxArrayString url;
    if (parameters->GetParamCount() > 0)
    {
        int i, nparams = parameters->GetParamCount();
        for (i = 0;i<nparams; i++)
            if (parameters->GetParam(i) != wxEmptyString)
                url.Add(parameters->GetParam(i));
    }
    if (parameters->Found(wxT("list"),&listtextfile))
    {
        wxLogNull nolog;
        wxTextFile file(listtextfile);
        if (file.Open())
        {
            wxString str;
            for ( str = file.GetFirstLine(); !file.Eof(); str = file.GetNextLine() )
                if (str != wxEmptyString)
                    url.Add(str);
            if (str != wxEmptyString)
                url.Add(str);
        }
    }
    if (mainframe->programoptions.rememberboxnewoptions)
    {
        numberofparts = mainframe->programoptions.lastnumberofparts;
        startoption = mainframe->programoptions.laststartoption;
        ontop = mainframe->programoptions.lastontopoption;
        if (!parameters->Found(wxT("destination"),&destination))
            destination = mainframe->programoptions.lastdestination;
        command = mainframe->programoptions.lastcommand;
    }
    else
    {
        command = wxEmptyString;
        numberofparts = DEFAULT_NUM_PARTS;
        startoption = DEFAULT_START_OPTION;
        ontop = DEFAULT_ONTOP_OPTION;
        if (!parameters->Found(wxT("destination"),&destination))
            destination = mainframe->programoptions.destination;
    }
    if (!parameters->Found(wxT("comments"),&comments))
        comments = wxEmptyString;
    if (!parameters->Found(wxT("reference"),&reference))
        reference = wxEmptyString;
    if (url.GetCount() > 0)
        mainframe->NewDownload(url,destination,-1,numberofparts,wxEmptyString,wxEmptyString,reference,comments,command,startoption,ontop,TRUE,TRUE);


    if (!parameters->Found(wxT("hide")))
        mainframe->Show(true);
    return true;
}

int mApplication::OnExit()
{
     // the mutex must be unlocked before being destroyed
     #ifndef DISABLE_MUTEX
     m_mutexAllDone.Unlock();
     #endif
//    delete m_checker;
    return 0;
}

wxString mApplication::Configurations(int operation, wxString option,wxString value)
{
    wxString tmpvalue = value;

    #ifdef WXDFAST_PORTABLE
    wxFileConfig *config = new wxFileConfig(DFAST_REG, wxEmptyString, DFAST_REG + wxT(".ini"), wxEmptyString,
                                                wxCONFIG_USE_LOCAL_FILE | wxCONFIG_USE_RELATIVE_PATH);
    #else
    wxFileConfig *config = new wxFileConfig(DFAST_REG);
    #endif

    config->SetPath(CONFIG_REG);
    if (operation == WRITE)
        config->Write(option,tmpvalue);
    else
        config->Read(option,&tmpvalue);
    delete config;
    return tmpvalue;
}

int mApplication::Configurations(int operation, wxString option,int value)
{
    int tmpvalue = value;
    #ifdef WXDFAST_PORTABLE
    wxFileConfig *config = new wxFileConfig(DFAST_REG, wxEmptyString, DFAST_REG + wxT(".ini"), wxEmptyString,
                                            wxCONFIG_USE_LOCAL_FILE | wxCONFIG_USE_RELATIVE_PATH);
    #else
    wxFileConfig *config = new wxFileConfig(DFAST_REG);
    #endif

    config->SetPath(CONFIG_REG);
    if (operation == WRITE)
        config->Write(option,tmpvalue);
    else
        config->Read(option,&tmpvalue);
    delete config;
    return tmpvalue;
}

long mApplication::Configurations(int operation, wxString option,long value)
{
    wxString tmpvalue;
    long returnvalue;
    tmpvalue << value;
    #ifdef WXDFAST_PORTABLE
    wxFileConfig *config = new wxFileConfig(DFAST_REG, wxEmptyString, DFAST_REG + wxT(".ini"), wxEmptyString,
                                            wxCONFIG_USE_LOCAL_FILE | wxCONFIG_USE_RELATIVE_PATH);
    #else
    wxFileConfig *config = new wxFileConfig(DFAST_REG);
    #endif

    config->SetPath(CONFIG_REG);
    if (operation == WRITE)
        config->Write(option,tmpvalue);
    else
        config->Read(option,&tmpvalue);
    delete config;
    tmpvalue.ToLong(&returnvalue);
    return returnvalue;
}

void mApplication::SetLanguage(int language)
{
    wxString msg;
    msg = _("Cannot set locale to");
    msg += wxT(" '") + m_locale->GetLanguageName(language) + wxT("'.\n");
    msg += _("This language is not supported by your system.");

    if (m_locale)
        delete m_locale;
    m_locale = new wxLocale();
    #if wxCHECK_VERSION(2, 8, 0)
    if (m_locale->IsAvailable(language))
        m_locale->Init(language);
    else
    {
        wxMessageBox(msg,_("Error...") ,wxOK | wxICON_ERROR,NULL);
        m_locale->Init(wxLANGUAGE_UNKNOWN);
    }
    #else
    m_locale->Init(language);
    #endif
    m_locale->AddCatalogLookupPathPrefix(wxT("languages"));
	m_locale->AddCatalogLookupPathPrefix(wxStandardPaths::Get().GetResourcesDir() + wxT("/locale"));
    m_locale->AddCatalog(wxT("wxDFast"));
}
